src/librustc/ty/structural_match.rs

```diff
@@ -207,3 +301,186 @@ impl<'a, 'tcx> TypeVisitor<'tcx> for Search<'a, 'tcx> {
         false
     }
 }
+
+struct SearchHirExpr<'a, 'tcx> {
```

Comment:
Not sure how I feel about this code being in `rustc::ty`, but that's neither here nor there.

What I might expect is to find this is in `rustc_mir::hair`, perhaps even operating on patterns extracted from the const expressions (to avoid having to reason about "expressions as patterns" in more than one place).

However, that might actually be a cyclic dependency of sorts, so potentially there's nothing to do here, for now.

Reply:
I'll look into this. (I had thought there was some motivation for making this accessible more broadly than would be possible if it lived in `rustc_mir`, but I should double-check that.)

Reply:
Actually, I just realized something - are we removing that `Expr` -> `Pat` conversion / was it already removed, on the basis that if we enforce `#[structural_match]` we can just use `==` instead?

Because then I think `mir_const_qualif` is probably the way to go, and then we don't have to look at HIR expressions and/or try to convert them into patterns.

Reply:
We have not removed the `Expr` -> `Pat` conversion, and we cannot use `==` instead in all cases, at least not today due to compiler bugs and/or language limitations.

The particular case I am thinking of is that we cannot implement a blanket `PartialEq` over all `for <'a> fn(...)`, but we *do* allow those to appear in patterns.

This leads to some cases where we ICE today, and other cases where we just rely on the `Expr` -> `Pat` conversion to deal with the codegen.

References:
 * https://github.com/rust-lang/rust/issues/46989#issuecomment-353805481
 * https://github.com/rust-lang/rust/issues/61188#issuecomment-539414038
 * https://github.com/rust-lang/rust/issues/65466